src/mpid/ch4/shm/ipc/src/ipc_pre.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef IPC_PRE_H_INCLUDED
+#define IPC_PRE_H_INCLUDED
+
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+#include "../xpmem/xpmem_pre.h"
+#endif
+
+typedef enum {
+    MPIDI_SHM_IPC_TYPE__NONE,   /* avoid empty enum */
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+    MPIDI_SHM_IPC_TYPE__XPMEM,
+#endif
+} MPIDI_SHM_IPC_type_t;
+
+typedef struct {
+    MPIR_Group *node_group_ptr; /* cache node group, used at win_create. */
+} MPIDI_IPC_global_t;
+
+/* memory handle definition
+ * MPIDI_IPC_mem_handle_t: local memory handle
+ * MPIDI_IPC_mem_seg_t: mapped segment with remote memory handle */
+typedef union {
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+    MPIDI_IPC_xpmem_mem_handle_t xpmem;
+#endif
+} MPIDI_IPC_mem_handle_t;
```

Comment:
Here is my understanding, @minsii correct me if I am wrong. At the moment, being this an XPMEM refactoring PR, we are keeping as much of XPMEM structure as possible unchanged. The AVL tree implementation was tailored for XPMEM and for this reason we are temporarily keeping it in XPMEM submodule. There is a plan to make a generic AVL tree structure in MPL so that GPU IPC submodule can also use it to cache IPC handles. When that happens we can refactor this part further. Please note that because of this, the GPU IPC submodule does not currently provide caching of memory handles. Does this make sense?

Reply:
If you notice my comment, I was careful not to suggest moving the AVL tree to the IPC layer.  I was saying that the address would be needed by all submodules uniformly, so it could be stored in the generic part of the structure, instead of the XPMEM part.  But this is a minor thing and it can be moved later too.

Reply:
Yes, it is correct. Kaiming is working on the MPL AVL tree generalization now (he will post a PR soon). Besides, we also want to understand whether AVL tree is better or HASH is better for the memory handle cache, especially for GPU.

@pavanbalaji Why do you cache the memory address for GPU? Should we cache the IPC handle `MPL_gpu_ipc_mem_handle_t`?

Reply:
How will the IPC handle work if I have an offset?  For example, address `0x100` (length `100 bytes`) might be cached, but I want to send `25` bytes from address `0x125`.  The memory handle itself is not useful to check for caching because the cache works on an address range.

Reply:
In the GPU PR, the sender always gets the IPC handle of sbuf when sending a message. We understand that the GPU driver has a similar address->IPC handle cache mechanism in order to ensure getting the handle is cheap. Then, on the receiver, it always receives the IPC handle and opens it. Therefore, it needs to check if the same `handle` has been cached. @gcongiu please correct me if I am wrong.

Reply:
I thought we wanted to cache the IPC handle as it is cheap to get the handle and costly to map it, which would be the case when using the virtual address for lookup. From what I could see, at least for CUDA, the IPC handle won't change if you query it for a subrange of a buffer. It will change only if you reallocate the buffer itself. So from that point of view, caching the IPC handle is even more reliable than using virtual address as for this it is possible to reallocate the buffer and get the same virtual address.

Reply:
There are a few incorrect points in your comment.  First, getting the IPC handle is not fast because of a cache.  It's just a fast function, whose speed doesn't rely on reuse.  Second, the caching is for opening the IPC handle.  That's the expensive part, and if I don't do caching, I might end up with one IPC handle covering a subset of another IPC handle.  There's no way for us to do range caching using IPC handles.  We'd need the actual address for that.

Reply:
@gcongiu My understanding is that the IPC handle encodes the address too.  So I don't think it can be the same handle for subrange buffers.

Reply:
Perhaps this discussion can be taken offline?